### PR TITLE
HEEDLS-432 - Centre Course Setup search sort filter and paginate impl…

### PR DESCRIPTION
**Pull request by DanBloxham-sw**
...emented

### JIRA link
This is split over three tickets:
Search & Paginate: https://softwiretech.atlassian.net/browse/HEEDLS-432
Sort: https://softwiretech.atlassian.net/browse/HEEDLS-433
Filter: https://softwiretech.atlassian.net/browse/HEEDLS-434

### Description
Implemented both JS and non-JS versions of searching, pagination, sorting, and filtering to the Course Setup page.

### Screenshots
JS UI with full page:
Desktop js search (2)

Non JS UI:
Desktop no js

Mobile:

-----
### Developer checks
I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (controller, data services, services, view models etc) and manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [x] Updated/added documentation in Swiki and/or Readme.
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/CourseSetup/CourseSetupControllerTests.cs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            new Category { CategoryName = "Category 1" },
+            new Category { CategoryName = "Category 2" }
+        };
+
+        private readonly List<CourseStatistics> courses = new List<CourseStatistics>
+        {
+            new CourseStatistics
+            {
+                ApplicationName = "Course", CustomisationName = "Customisation", Active = true, CourseTopic = "Topic 1",
+                CategoryName = "Category 1", HideInLearnerPortal = true, DelegateCount = 1, CompletedCount = 1
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/CourseSetup/CourseSetupControllerTests.cs:30]: This initialiser shouldn't be formatted like this

File: DigitalLearningSolutions.Web/Controllers/TrackingSystem/CourseSetup/CourseSetupController.cs [hunk 2]
```diff
@@ -13,22 +18,99 @@
...
     public class CourseSetupController : Controller
     {
+        private const string CourseFilterCookieName = "CourseFilter";
+        private static readonly DateTimeOffset CookieExpiry = DateTimeOffset.UtcNow.AddDays(30);
+        private readonly ICourseCategoriesDataService courseCategoriesDataService;
+        private readonly ICourseTopicsDataService courseTopicsDataService;
         private readonly ICourseService courseService;
 
-        public CourseSetupController(ICourseService courseService)
+        public CourseSetupController(ICourseService courseService, ICourseCategoriesDataService courseCategoriesDataService, ICourseTopicsDataService courseTopicsDataService)
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web/Controllers/TrackingSystem/CourseSetup/CourseSetupController.cs:27]: Long line, run formatter

File: DigitalLearningSolutions.Web/Controllers/TrackingSystem/CourseSetup/CourseSetupController.cs [hunk 3]
```diff
@@ -13,22 +18,99 @@
...
             var centreCourses = courseService.GetCentreSpecificCourseStatistics(centreId, categoryId.Value);
-            var model = new CourseSetupViewModel(centreCourses.Take(10));
+            var categories = courseCategoriesDataService.GetCategoriesForCentreAndCentrallyManagedCourses(centreId).Select(c => c.CategoryName);
+            var topics = courseTopicsDataService.GetCourseTopicsAvailableAtCentre(centreId).Select(c => c.CourseTopic);
 
+            var model = new AllCourseStatisticsViewModel(centreCourses, categories, topics);
             return View(model);
         }
+
+        private void UpdateFilterCookie(string? filterBy)
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web/Controllers/TrackingSystem/CourseSetup/CourseSetupController.cs:97]: I took a slightly different approach to this comment and only took out the contents of the true branch to a new method. We should be consistent, but I'm not sure which way.

If we take the approach here, the method needs renaming to something referencing the fact that it may also be deleted here.

**Review by AlexJacksonDS — COMMENTED** [review]
Just a few minor things, other than those it looks OK to me

**Review by AlexJacksonDS — APPROVED** [review]

File: DigitalLearningSolutions.Data/DataServices/CourseTopicsDataService.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            this.connection = connection;
+        }
+
+        public IEnumerable<Topic> GetCourseTopicsAvailableAtCentre(int centreId)
+        {
+            return connection.Query<Topic>(
+                @"SELECT CourseTopicID, CourseTopic, Active
+                    FROM CourseTopics
+                    WHERE ((CentreID = @CentreID) OR (CentreID = 0)) AND (Active = 1)
```
Comment (stellake) [inline DigitalLearningSolutions.Data/DataServices/CourseTopicsDataService.cs:27]: Do we need brackets around (CentreID = @CentreID) and (CentreID = 0)

File: DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/CourseSetup/CourseSetupControllerTests.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+            // TODO: when 507 is merged, replace this with the new .WithMockHttpContextWithCookie controller extension method
+            httpContext = A.Fake<HttpContext>();
+            httpRequest = A.Fake<HttpRequest>();
+            httpResponse = A.Fake<HttpResponse>();
+            cookieCollection = A.Fake<IRequestCookieCollection>();
+
+            const string cookieName = "CourseFilter";
+            const string cookieValue = "Status|Active|true";
+            var cookieList = new List<KeyValuePair<string, string>>
```
Comment (stellake) [inline DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/CourseSetup/CourseSetupControllerTests.cs]: Will use the helper after Alex's merge
Will also need to check the filterBy changes

File: DigitalLearningSolutions.Web/Scripts/spec/sort.spec.ts [hunk 6]
```diff
@@ -286,6 +286,62 @@ describe('sortSearchableElements delegates', () => {
   });
 });
 
+describe('sortSearchableElements delegates groups', () => {
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Scripts/spec/sort.spec.ts]: How is this related to the ticket?

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/CourseSetup/AllCourseStatisticsViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    using System.Collections.Generic;
+    using System.Linq;
+    using DigitalLearningSolutions.Data.Models.Courses;
+    using DigitalLearningSolutions.Web.ViewModels.Common.SearchablePage;
+
+    public class AllCourseStatisticsViewModel : BaseJavaScriptFilterableViewModel
+    {
+        public AllCourseStatisticsViewModel(IEnumerable<CourseStatistics> courses, IEnumerable<string> categories, IEnumerable<string> topics)
+        {
+            courses = courses.ToList();
```
Comment (stellake) [inline DigitalLearningSolutions.Web/ViewModels/TrackingSystem/CourseSetup/AllCourseStatisticsViewModel.cs:12]: Why are we doing .ToList() instead of changing the param type to more specific?

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/CourseSetup/AllCourseStatisticsViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+            Courses = courses.Select(c => new SearchableCourseStatisticsViewModel(c));
+
+            Filters = CourseStatisticsViewModelFilterOptions.GetFilterOptions(categories, topics)
+                .Select(
+                    f => f.FilterOptions.Select(
+                        fo => new AppliedFilterViewModel(fo.DisplayText, f.FilterName, fo.FilterValue)
+                    )
+                ).SelectMany(af => af).Distinct();
+        }
```
Comment (stellake) [inline DigitalLearningSolutions.Web/ViewModels/TrackingSystem/CourseSetup/AllCourseStatisticsViewModel.cs:21]: I've seen this in many places - should this be commonised somehow?

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/CourseSetup/CourseSetupViewModel.cs [hunk 9]
```diff
@@ -3,14 +3,47 @@
...
+            IEnumerable<string> topics,
+            string? searchString,
+            string sortBy,
+            string sortDirection,
+            string? filterBy,
+            int page
+        ) : base(searchString, page, true, sortBy, sortDirection, filterBy)
         {
-            Courses = courses.Select(course => new SearchableCourseStatisticsViewModel(course));
+            courses = courses.ToList();
```
Comment (stellake) [inline DigitalLearningSolutions.Web/ViewModels/TrackingSystem/CourseSetup/CourseSetupViewModel.cs]: Is there any reason of doing IEnumerable<CourseStatistics> type courses as a param, then override it with .ToList(), then change it into Queryable, then never use it again?

File: DigitalLearningSolutions.Web/Views/TrackingSystem/CourseSetup/Index.cshtml [hunk 10]
```diff
@@ -19,15 +19,42 @@
   <div class="nhsuk-grid-column-full">
     <h1 class="nhsuk-heading-xl">Centre course setup</h1>
 
+    <div class="nhsuk-grid-row">
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Views/TrackingSystem/CourseSetup/Index.cshtml:22]: Is column already in the VC?

File: DigitalLearningSolutions.Web/Scripts/spec/sort.spec.ts (hunk 6 above)
Comment (stellake) [reply DigitalLearningSolutions.Web/Scripts/spec/sort.spec.ts]: Oh I see - we should probably change the name of this test

File: DigitalLearningSolutions.Web/Views/TrackingSystem/CourseSetup/Index.cshtml (hunk 10 above)
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Web/Views/TrackingSystem/CourseSetup/Index.cshtml:22]: Yep, both the search-box VC and Sort have the column div inside. The following two have neither, hence why the row and column is present in the Index page.

Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Views/TrackingSystem/CourseSetup/Index.cshtml:22]: This is one of the things I noted to fix when fixing all the styling issues.

Comment (stellake) [reply DigitalLearningSolutions.Web/Views/TrackingSystem/CourseSetup/Index.cshtml:22]: Happy as long as it's documented on the ticket 👌

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/CourseSetup/CourseStatisticsViewModelFilterOptions.cs [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+                    topic,
+                    nameof(CourseStatistics.CourseTopic) + FilteringHelper.Separator +
+                    nameof(CourseStatistics.CourseTopic) +
+                    FilteringHelper.Separator + topic,
+                    FilterStatus.Default
+                )
+            );
+        }
+
+        public static IEnumerable<FilterViewModel> GetFilterOptions(
```
Comment (stellake) [inline DigitalLearningSolutions.Web/ViewModels/TrackingSystem/CourseSetup/CourseStatisticsViewModelFilterOptions.cs:51]: super minor, but public methods usually come before private methods in the file

File: DigitalLearningSolutions.Web/Controllers/TrackingSystem/CourseSetup/CourseSetupController.cs [hunk 12]
```diff
@@ -13,21 +16,83 @@
...
+        )
         {
+            if (filterBy == null && filterValue == null)
+            {
+                filterBy = Request.Cookies[CourseFilterCookieName];
+                if (filterBy == null)
+                {
+                    // Only show Active courses by default
+                    filterValue = CourseStatusFilterOptions.IsActive.FilterValue;
+                }
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Controllers/TrackingSystem/CourseSetup/CourseSetupController.cs:52]: Should this just be `filterBy = Request.Cookies[CourseFilterCookieName] ?? CourseStatusFilterOptions.IsActive.FilterValue`?

File: DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/CourseSetup/CourseSetupControllerTests.cs [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+            var result = controller.Index(filterBy: filterBy);
+
+            // Then
+            A.CallTo(() => httpResponse.Cookies.Delete("CourseFilter")).MustHaveHappened();
+            result.As<ViewResult>().Model.As<CourseSetupViewModel>().FilterBy.Should()
+                .BeNull();
+        }
+
+        [Test]
+        public void Index_with_null_filterBy_and_new_filter_query_parameter_add_new_cookie_value()
```
Comment (stellake) [inline DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/CourseSetup/CourseSetupControllerTests.cs:118]: adds?

File: DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/CourseSetup/CourseSetupControllerTests.cs [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+            A.CallTo(() => courseService.GetCentreSpecificCourseStatistics(A<int>._, A<int>._)).Returns(courses);
+            A.CallTo(() => courseCategoryDataService.GetCategoriesForCentreAndCentrallyManagedCourses(A<int>._))
+                .Returns(categories);
+            A.CallTo(() => courseTopicsDataService.GetCourseTopicsAvailableAtCentre(A<int>._)).Returns(topics);
+
+            httpRequest = A.Fake<HttpRequest>();
+            httpResponse = A.Fake<HttpResponse>();
+            const string cookieName = "CourseFilter";
+            const string cookieValue = "Status|Active|true";
```
Comment (stellake) [inline DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/CourseSetup/CourseSetupControllerTests.cs:68]: I feel like this cookie value is not an optimal choice as we will probably want a test for the case of "no cookie, status is set to active" case. It would be good if the filter in the cookie is different from that test case.

File: DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/CourseSetup/CourseSetupControllerTests.cs [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+            // Then
+            result.As<ViewResult>().Model.As<CourseSetupViewModel>().FilterBy.Should()
+                .Be(filterBy);
+        }
+
+        [Test]
+        public void Index_with_CLEAR_filterBy_query_parameter_removes_cookie()
+        {
+            // Given
+            const string? filterBy = "CLEAR";
```
Comment (stellake) [inline DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/CourseSetup/CourseSetupControllerTests.cs:106]: super minor, but why is this type nullable if it's const?

**Review by stellake — APPROVED** [review]
Looks good! Could you please double check before merge that the VC inconsistency is mentioned in JIRA? 🙏
